futures-02.md

```diff
@@ -0,0 +1,396 @@
+    unsafe fn initializer(&self) -> Initializer { ... }
+
+    fn poll_read(&mut self, buf: &mut [u8], ctx: &mut Context)
+        -> Poll<usize, Self::Error>;
+
+    fn poll_vectored_read(&mut self, vec: &mut [&mut IoVec], ctx: &mut Context)
+        -> Poll<usize, Self::Error> { ... }
+}
+
+pub trait AsyncWrite {
+    type Error;
+    fn poll_write(&mut self, buf: &[u8], ctx: &mut Context)
+        -> Poll<usize, Self::Error>;
+
+    fn poll_vectored_write(&mut self, vec: &[&IoVec], ctx: &mut Context)
+        -> Poll<usize, Self::Error> { ... }
+
+    fn poll_close(&mut self, ctx: &mut Context) -> Poll<(), Self::Error>;
+}
+```
+
+In terms of the old `futures` task model, the `Waker` is analogous to the `Task`
+returned by [`task::current`][current], and the `wake` method is analogous to
+the old [`Task::notify`][notify].
+
+Astute readers will notice other changes in the traits shown above.
+These will be addressed in later sections.
+
+[current]: https://docs.rs/futures/0.1.16/futures/task/fn.current.html
+[notify]: https://docs.rs/futures/0.1.16/futures/task/struct.Task.html#method.notify
+
+## Crate Separation
+
+The `futures` crate will be split into several different libraries.
+
+### `futures-core`
+
+`futures-core` will include `Future`, `Stream`, `IntoFuture`
+traits and their dependencies. This is a central crate that includes only
+essential items, designed to minimize future breakage. Combinator methods
+will be removed from the `Future` and `Stream` and moved to the
+`FutureExt` and `StreamExt` traits in the `futures-util` crate. `poll` 
+will be the only remaining method on the `Future` and `Stream` traits.
+This separation allows combinator functions to evolve over time without
+breaking changes to the public API surface of async libraries.
+
+### `futures-io`
+
+`futures-io` includes the updated `AsyncRead` and `AsyncWrite` traits.
+These traits were previously included in the `tokio-io` crate.
+However, `AsyncRead` and `AsyncWrite` don't have any dependencies on the `tokio`
+crate or the rest of the `tokio` stack, so they will be moved into the `futures`
+group of libraries. `futures-io` depends only on the `futures-core` crate
+and the `bytes` crate (for the [`Buf`] and [`BufMut`] traits).
+
+[`Buf`]: https://carllerche.github.io/bytes/bytes/trait.Buf.html
+[`BufMut]: https://carllerche.github.io/bytes/bytes/trait.BufMut.html
+
+### `futures-sink`
+
+`futures-sink` includes the `Sink` trait, which allows values to be sent
+asynchronously. The `Sink` trait is excluded from `futures-core` because its
+design is much less stable than `Future` and `Stream`. There are several
+potential breaking changes to the `Sink` trait under consideration.
+
+This crate will depend only on `futures-core`.
+
+### `futures-channel`
+
+`futures-channel` includes the `mpsc` and `oneshot` messaging primitives.
+This crate is a minimal addition on `futures-core` designed for low churn
+so that channels may appear in the public APIs of libraries.
```

Comment:
But it will depend on `futures-sink` which is going to be less stable, right?

Reply:
I had actually imagined this as being inverted-- `futures-sink` would depend on `futures-channel` and provide the implementation there, since `futures-sink` is less stable than `futures-channel`.

Reply:
Would `Sender` than have a `poll_send` inherent method then?

Reply:
Yeah, that was my plan, but I'm open to suggestions :). If you think that seems reasonable, I'll add a bit about this to the RFC to explain.

Reply:
Yeah that seems reasonable, but I think it'd be good to clarify :)

Reply:
I've added a clarification to the RFC.